http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dns/JupyterNameService.scala

```diff
@@ -25,3 +27,8 @@ class JupyterNameServiceDescriptor extends NameServiceDescriptor {
   override def getProviderName: String = "Jupyter"
   override def getType: String = "dns"
 }
+
+object HostToIpMapping {
+  private[dns] val hostToIpMapping: ValueBox[Map[Host, IP]] = ValueBox(Map.empty)
```

Comment:
what does `ValueBox` do?

Reply:
So we need to update this global, static variable `hostToIpMapping` to be able to plug into Java DNS stuff. `ValueBox` provides some synchronization around updating this. I think originally we just used `@volatile` but we need to do a get-and-update so extra synchronization is needed. I think @kyuksel originally added `ValueBox`.

It'd be nice to use `IO` or maybe `Ref` instead but I recall this being hard to plug into `sun.net.spi.nameservice.NameService` which is some magic Java stuff (but maybe it's possible).

Reply:
maybe try `Ref[IO, Map[Host, IP]]`? https://typelevel.org/cats-effect/concurrency/ref.html

Reply:
Yeah.. maybe will try that

Reply:
Using `Ref[IO, Map[Host, IP]]` in my latest commit.. it seems to work